js/src/json/match.ts

```diff
@@ -0,0 +1,466 @@
+    let processedReferenceOutputs: Record<string, any> = {};
+
+    if (Array.isArray(outputs)) {
+        useListReducer = true;
+        if (!Array.isArray(referenceOutputs)) {
+            throw new Error('If outputs is a list, referenceOutputs must also be a list');
+        }
+
+        let outputsToUse: Record<string, any> = {};
+        let referenceOutputsToUse: Record<string, any> = {};
+
+        if (listMatchMode === 'ordered') {
+            outputs.forEach((output, i) => {
+                Object.entries(output).forEach(([key, value]) => {
+                    outputsToUse[`${key}_${i}`] = value;
+                });
+            });
+            referenceOutputs.forEach((refOutput, i) => {
+                Object.entries(refOutput).forEach(([key, value]) => {
+                    referenceOutputsToUse[`${key}_${i}`] = value;
+                });
+            });
+        } else if (listMatchMode === 'superset') {
+            const availableOutputs = Array.from(Array(outputs.length).keys());
+            const matchedReferences = new Set<number>();
+
+            referenceOutputs.forEach((refItem: any, i: number) => {
+                let bestMatchScore = -1;
+                let bestMatchIdx: number | null = null;
+
+                availableOutputs.forEach(outIdx => {
+                    const outputItem = outputs[outIdx];
+                    let matchScore = 0;
+
+                    Object.keys(refItem).forEach(key => {
+                        if (key in outputItem && !excludeKeys.includes(key) && !(key in rubric)) {
+                            matchScore += Number(refItem[key] === outputItem[key]);
+                        }
+                    });
+
+                    if (matchScore > bestMatchScore) {
+                        bestMatchScore = matchScore;
+                        bestMatchIdx = outIdx;
+                    }
+                });
+
+                if (bestMatchIdx !== null) {
+                    Object.entries(outputs[bestMatchIdx]).forEach(([key, value]) => {
+                        outputsToUse[`${key}_${i}`] = value;
+                    });
+
+                    Object.entries(refItem).forEach(([key, value]) => {
+                        referenceOutputsToUse[`${key}_${i}`] = value;
+                    });
+
+                    availableOutputs.splice(availableOutputs.indexOf(bestMatchIdx), 1);
+                    matchedReferences.add(i);
+                } else {
+                    Object.entries(refItem).forEach(([key, value]) => {
+                        referenceOutputsToUse[`${key}_${i}`] = value;
+                    });
+                }
+            });
+        } else {
+            const availableReferences = Array.from(Array(referenceOutputs.length).keys());
+            const matchedOutputs = new Set<number>();
+
+            outputs.forEach((outputItem: any, i: number) => {
+                let bestMatchIdx: number | null = null;
```

Comment:
same nit as above on types

```suggestion
            outputs.forEach((outputItem, i) => {
                let bestMatchIdx: number | undefined;
```